File: 2958-Length-of-Longest-Subarray-With-at-Most-K-Frequency.js
var maxSubarrayLength = function(nums, k) {
    const n = nums.length;
    let longestGood = 0;
    let j = -1;
    let numCounter = {};

    for (let i = 0; i < n; i++) {
        let num = nums[i];
        numCounter[num] = (numCounter[num] || 0) + 1;
        while (numCounter[num] > k) {
            j += 1;
            numCounter[nums[j]] -= 1;
        }
        longestGood = Math.max(longestGood, i - j);
    }

    return longestGood;
};

File: 2958-Length-of-Longest-Subarray-With-at-Most-K-Frequency.py
class Solution(object):
    def maxSubarrayLength(self, nums, k):
        numCounter = Counter()
        j = -1
        longestGood = 0

        for i in range(len(nums)):
            num = nums[i]
            numCounter[num] += 1
            while numCounter[num] > k:
                j += 1
                numCounter[nums[j]] -= 1
            longestGood = max(longestGood, i - j)

        return longestGood
